src/main/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/onetimecode/OneTimeCodeService.kt

```diff
@@ -19,34 +25,44 @@ class OneTimeCodeService(
+      .also { oneTimeCodeAttemptsRepository.save(OneTimeCodeAttempts(sessionId, emptySet())) }
       .also { oneTimeCode -> oneTimeCodeEmailSender.send(email, oneTimeCode.code) }
   }
 
   fun verifyOneTimeCode(code: String, sessionId: String): String =
     if (code == smokeTestConfig.lsjSecret) {
       smokeTest()
     } else {
-      // Initial MVP implementation that does not support retries
-      oneTimeCodeRepository.findById(sessionId)
-        .orElseGet { throw ResourceNotFoundException("One Time Code not found") } // OTC for the specified session ID not found
-        .also { oneTimeCodeRepository.deleteById(sessionId) }
-        .takeIf { oneTimeCode -> oneTimeCode.code.equals(code, ignoreCase = true) }
-        ?.let { oneTimeCode -> jwtService.generateToken(oneTimeCode.email, findOrganisation(oneTimeCode.email)) }
-        ?: throw ResourceNotFoundException("One Time Code not found") // OTC for the session ID found, but not matching the specified code
-
-      /*
-      // Implementation that will be the basis of a 3-strikes-and-out retry solution, in that the OTC is not deleted immediately if the code does not match.
-      // However, because no count or state for 3-strikes-and-out is currently in place, codes do not get deleted, and can be repeatedly retried up until
-      // their 30 minute expiration. Likely to be addressed in SLM-233
-      oneTimeCodeRepository.findById(sessionId)
-        .orElseGet { throw ResourceNotFoundException("One Time Code not found") }
-        .takeIf { oneTimeCode -> oneTimeCode.code.equals(code, ignoreCase = true) }
-        ?.also { oneTimeCodeRepository.deleteById(sessionId) }
-        ?.let { oneTimeCode -> jwtService.generateToken(oneTimeCode.email, findOrganisation(oneTimeCode.email)) }
-        ?: throw ResourceNotFoundException("One Time Code not found") // TODO - this is where '3 strikes and out' will be handled when we get to that
-       */
+
+      val oneTimeCode = oneTimeCodeRepository.findById(sessionId).toNullable()
+      val attempts = oneTimeCodeAttemptsRepository.findById(sessionId).toNullable()
+        ?.let {
+          oneTimeCodeAttemptsRepository.save(
+            OneTimeCodeAttempts(sessionId, it.attempts?.plus(code.uppercase()) ?: mutableSetOf(code.uppercase()))
+          )
+        }
+      when {
+        oneTimeCode == null || attempts == null -> {
+          clearOneTimeCode(sessionId)
+          throw OneTimeCodeException(OneTimeCodeSessionNotFound)
+        }
+        oneTimeCode.code.equals(code, ignoreCase = true) -> {
+          clearOneTimeCode(sessionId)
+          jwtService.generateToken(oneTimeCode.email, findOrganisation(oneTimeCode.email))
+        }
+        attempts.attempts!!.size >= oneTimeCodeConfig.maxAttempts -> {
+          clearOneTimeCode(sessionId)
+          throw OneTimeCodeException(OneTimeCodeTooManyAttempts)
+        }
+        else -> throw OneTimeCodeException(OneTimeCodeNotFound)
```

Comment:
So `OneTimeCodeNotFound` is used when there is a OTC for the session in redis, and its attempts count has not yet been met, but the code provided by the user is not the code of the OTC ?
Is that right?

Reply:
Yes. It's a straight replacement for the old 404 Not Found. In summary we have:
* OTC Session not found - we have no OTC to check against (because it has timed out)
* OTC Too many attempts - we have guessed the OTC incorrectly too many times
* OTC Not found - the last guess was incorrect and the above don't apply